File: dsa/concepts/13_graphs/java/TopoSort_L269_AlienDictionary.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.Stack;

public class TopoSort_L269_AlienDictionary {

	public static void main(String[] args) {
		TopoSort_L269_AlienDictionary obj = new TopoSort_L269_AlienDictionary();

		String[] words = new String[] {"wrt","wrf","er","ett","rftt"};
		// String[] words = new String[] {"zy","zx"};
		System.out.println(obj.alienOrder(words));
	}

	// Problem: https://www.lintcode.com/problem/892
	// Idea: Words are already sorted lexicographically by the rules of this new language. Hence, we can find the relationship between the chars from adjecent words.
	// Build a relationship graph from adjacent word chars.
	// From this graph reverse topological order is tha answer
	public String alienOrder(String[] words) {
		Map<Character, List<Character>> adj = new HashMap<>();
		for(String word: words) {
			for(char ch: word.toCharArray()) {
				adj.putIfAbsent(ch, new ArrayList<>());
			}
		}

		for(int i=0; i<words.length-1; i++) {
			String w1 = words[i], w2 = words[i+1];
			int min = Math.min(w1.length(), w2.length());
			for(int j=0; j<min; j++) {
				if(w1.charAt(j) != w2.charAt(j)) {
					adj.get(w1.charAt(j)).add(w2.charAt(j));
					break;		// we have to break as we are sure only first chars are in lex order.
				}
			}
		}

		Set<Character> vis = new HashSet<>();
		Set<Character> pathVis = new HashSet<>();
		Stack<Character> st = new Stack<>();
		for(char ch: adj.keySet()) {
			if(!vis.contains(ch)){
				if(!dfs(adj, ch, st, vis, pathVis))
					return "";
			}
		}
		StringBuffer sb = new StringBuffer();
		while(!st.isEmpty()) sb.append(st.pop());
		return sb.toString();
	}

	private boolean dfs(Map<Character, List<Character>> adj, char ch, Stack<Character> st,
			Set<Character> vis,	Set<Character> pathVis) {
		vis.add(ch);
		pathVis.add(ch);
		for(char nei: adj.get(ch)) {
			if(!vis.contains(nei)) {
				if(!dfs(adj, nei, st, vis, pathVis))
					return false;
			}else if(pathVis.contains(nei)) {
				return false;
			}
		}
		pathVis.remove(ch);
		st.add(ch);
		return true;
	}

}

File: dsa/topics/13_graphs/python/topo_sort_l269_alien_dictionary.py
from typing import List, Dict


"""
Problem:
https://www.lintcode.com/problem/892
https://leetcode.com/problems/alien-dictionary/

Idea:
Words are already sorted lexicographically according to an unknown alien language.

Approach:
1. Build a graph of character precedence from adjacent words
2. For each adjacent word pair, find the first differing character
   and add an edge: char1 -> char2
3. Perform DFS-based topological sort on characters
4. Use pathVis to detect cycles (invalid ordering)
5. Reverse postorder gives alien dictionary order

Time Complexity: O(C + E)
Space Complexity: O(C + E)
Where C = number of unique characters
"""
def alien_order(words: List[str]) -> str:
    adj: Dict[str, List[str]] = {}

    # Initialize graph with all unique characters
    for word in words:
        for ch in word:
            if ch not in adj:
                adj[ch] = []

    # Build precedence graph from adjacent words
    for i in range(len(words) - 1):
        w1, w2 = words[i], words[i + 1]
        min_len = min(len(w1), len(w2))

        for j in range(min_len):
            if w1[j] != w2[j]:
                adj[w1[j]].append(w2[j])
                break

    vis = set()
    path_vis = set()
    stack = []

    for ch in adj:
        if ch not in vis:
            if not dfs(adj, ch, vis, path_vis, stack):
                return ""

    return "".join(reversed(stack))


"""
DFS helper for topological sorting with cycle detection

Returns False if a cycle is detected
"""
def dfs(adj: Dict[str, List[str]], ch: str,
        vis: set, path_vis: set, stack: List[str]) -> bool:
    vis.add(ch)
    path_vis.add(ch)

    for nei in adj[ch]:
        if nei not in vis:
            if not dfs(adj, nei, vis, path_vis, stack):
                return False
        elif nei in path_vis:
            return False

    path_vis.remove(ch)
    stack.append(ch)
    return True


# one test (same as Java main)
if __name__ == "__main__":
    words = ["wrt", "wrf", "er", "ett", "rftt"]
    print(alien_order(words))
